lib/auth/saml.go

```diff
@@ -336,87 +406,131 @@ func (a *Server) ValidateSAMLResponse(samlResponse string) (*SAMLAuthResponse, e
+		traceErr("Failed to retrieve assertion info. This may indicate IdP configuration error.", err)
+		return nil, trace.Wrap(err)
 	}
 
+	diagInfo(types.SSOInfoType_SAML_ASSERTION_INFO, assertionInfo)
+
 	if assertionInfo.WarningInfo.InvalidTime {
-		return nil, trace.AccessDenied("invalid time in SAML assertion info")
+		err = trace.AccessDenied("invalid time in SAML assertion info")
+		traceErr("SAML assertion info contained warning: invalid time.", err)
+		return nil, trace.Wrap(err)
 	}
 
 	if assertionInfo.WarningInfo.NotInAudience {
-		return nil, trace.AccessDenied("no audience in SAML assertion info")
+		err = trace.AccessDenied("no audience in SAML assertion info")
+		traceErr("SAML: not in expected audience. Check auth connector audience field and IdP configuration for typos and other errors.", err)
+		return nil, trace.Wrap(err)
 	}
 
 	log.Debugf("Obtained SAML assertions for %q.", assertionInfo.NameID)
-	re := &samlAuthResponse{
-		attributeStatements: make(map[string][]string),
-	}
+
+	attributeStatements := map[string][]string{}
+
 	for key, val := range assertionInfo.Values {
 		var vals []string
 		for _, vv := range val.Values {
 			vals = append(vals, vv.Value)
 		}
 		log.Debugf("SAML assertion: %q: %q.", key, vals)
-		re.attributeStatements[key] = vals
+		attributeStatements[key] = vals
 	}
 
+	auxInfo.attributeStatements = attributeStatements
+
+	diagInfo(types.SSOInfoType_SAML_ATTRIBUTE_STATEMENTS, attributeStatements)
+
 	log.Debugf("SAML assertion warnings: %+v.", assertionInfo.WarningInfo)
 
+	diagInfo(types.SSOInfoType_SAML_ATTRIBUTES_TO_ROLES, connector.GetAttributesToRoles())
+
 	if len(connector.GetAttributesToRoles()) == 0 {
-		return re, trace.BadParameter("no attributes to roles mapping, check connector documentation")
+		err = trace.BadParameter("no attributes to roles mapping, check connector documentation")
+		traceErr("Attributes-to-roles mapping is empty, SSO user will never have any roles.", err)
+		return nil, trace.Wrap(err)
 	}
+
 	log.Debugf("Applying %v SAML attribute to roles mappings.", len(connector.GetAttributesToRoles()))
 
 	// Calculate (figure out name, roles, traits, session TTL) of user and
 	// create the user in the backend.
-	params, err := a.calculateSAMLUser(connector, *assertionInfo, request)
+	params, err := a.calculateSAMLUser(ctx, connector, *assertionInfo, request)
 	if err != nil {
-		return re, trace.Wrap(err)
+		traceErr("Failed to calculate user attributes.", err)
+		return nil, trace.Wrap(err)
 	}
-	user, err := a.createSAMLUser(params)
+
+	diagInfo(types.SSOInfoType_CREATE_USER_PARAMS, params)
+
+	user, err := a.createSAMLUser(params, request.SSOTestFlow)
 	if err != nil {
-		return re, trace.Wrap(err)
+		traceErr("Failed to create user from provided parameters.", err)
+		return nil, trace.Wrap(err)
```

Comment:
Instead of doing `traceErr` in each "err != nil" branch, can you update the returned errors to include user message and add the error to the diagnostic once at the caller side?

```go
return nil, trace.Wrap(err, "failed to create user from provider parameter")
```

Reply:
Done: e478da1c5ac88a1dbe5f21b4af20772dc980da82